src/redisjson.rs

```diff
@@ -81,20 +86,78 @@ impl RedisJSON {
     pub fn str_len(&self, path: &str) -> Result<usize, Error> {
         match self.get_doc(path)?.as_str() {
             Some(s) => Ok(s.len()),
-            None => Err(Error{msg: "ERR wrong type of path value".to_string()})
+            None => Err("ERR wrong type of path value".into())
+        }
+    }
+
+    pub fn arr_len(&self, path: &str) -> Result<usize, Error> {
+        match self.get_doc(path)?.as_array() {
+            Some(s) => Ok(s.len()),
+            None => Err("ERR wrong type of path value".into())
+        }
+    }
+
+    pub fn obj_len(&self, path: &str) -> Result<usize, Error> {
+        match self.get_doc(path)?.as_object() {
+            Some(s) => Ok(s.len()),
+            None => Err("ERR wrong type of path value".into())
         }
     }
 
     pub fn get_type(&self, path: &str) -> Result<String, Error> {
-        let s = match self.get_doc(path)? {
+        let s = RedisJSON::value_name(self.get_doc(path)?);
+        Ok(s.to_string())
+    }
+
+    fn value_name(value: &Value) -> &str {
+        match value {
             Value::Null => "null",
             Value::Bool(_) => "boolean",
             Value::Number(_) => "number",
             Value::String(_) => "string",
             Value::Array(_) => "array",
             Value::Object(_) => "object",
-        };
-        Ok(s.to_string())
+        }
+    }
+
+    pub fn num_op<F: Fn(f64, f64) -> f64>(&mut self, path: &str, number: f64, fun: F) -> Result<String, Error> {
```

Comment:
This is quite heavily nested with multiple `match`es and I found it hard to understand.

I also didn't like the `mut error`. It's not very Rust-like, since it uses a side effect of mutating instead of returning a pure value. Unfortunately, the `replace_with` method doesn't return `Result<Value, Error>` but just a `Value`, so it's very hard to do this in a Rust-like way :-(.

I cleaned it up a bit by using `if let` instead of `match` and by using a `Vec<String>` instead of concatenating strings.

I think I can improve this even further, but let's start with this for now:
```
    pub fn num_op<F: Fn(f64, f64) -> f64>(
        &mut self,
        path: &str,
        number: f64,
        fun: F,
    ) -> Result<String, Error> {
        let current_data = mem::replace(&mut self.data, Value::Null);

        let mut errors = vec![];
        let mut result: f64 = 0.0;

        self.data = jsonpath_lib::replace_with(current_data, path, &mut |v| {
            if let Value::Number(curr) = v {
                if let Some(curr_value) = curr.as_f64() {
                    result = fun(curr_value, number);

                    if let Some(new_value) = Number::from_f64(result) {
                        Value::Number(new_value)
                    } else {
                        errors.push("ERR can not represent result as Number".to_string());
                        v.clone()
                    }
                } else {
                    errors.push("ERR can not convert current value as f64".to_string());
                    v.clone()
                }
            } else {
                errors.push(format!("ERR wrong type of path value - expected a number but found {}",
                                       RedisJSON::value_name(&v)));
                v.clone()
            }
        })?;

        if errors.is_empty() {
            Ok(result.to_string())
        } else {
            Err(errors.join("\n").into())
        }
    }
```